Fix getting subnet of ipv4 mapped ipv6 addresses

PVince81 opened the pull request:
:warning: my knowledge of IPv6 is very limited, so please help double checking whether this makes sense at all

PVince81 committed:
Fix getting subnet of ipv4 mapped ipv6 addresses

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

nickvergessen commented on lib/private/Security/Normalizer/IpAddress.php:
```diff
@@ -103,6 +103,8 @@ public function getSubnet(): string {
 				$this->ip,
 				32
 			);
+		} else if (substr(strtolower($this->ip), 0, 7) === '::ffff:') {
```
```suggestion
		} elseif (substr(strtolower($this->ip), 0, 7) === '::ffff:') {
```

Does this always mean there is an IPv4 inside? Or can it also be another ipv6?

PVince81 replied:
I found this in https://en.wikipedia.org/wiki/IPv6_address#Transition_from_IPv4 but the text doesn't fully match
also https://en.wikipedia.org/wiki/IPv6_address#Special_addresses

I can't say for sure but the more I read about this the more confused I am...

juliushaertl replied:
Yes, last 32 bits of the ipv6 address are considered to be interpreted as IPv4 (if prefix matches ::ffff:)

Basically the whole IPv4 adress space is then mapped into `::ffff:0:0/96`. The thing I'm not sure about here is, if expanded notation as in `0000:0000:0000:0000:0000:ffff:7f00:0001` should also taken into consideration, which would be the same as `::ffff:127.0.0.1`

come-nc replied:
If I understand correctly the code you cannot know for sure the IP will be in canonical form yet?
How does the test behave if you add something like ['::0:ffff:192.168.0.123','::ffff:192.168.0.123/32'] ?

Otherwise from what I understand this prefix is reserved to IPv4 mapped addresses.

PVince81 replied:
yeah, probably I'll use the regexp instead and limit it strictly to "::ffff:x.x.x.x" format ?
or how many more formats do we need to add there ?

PVince81 replied:
(also wondering why we're not using a library instead of writing own code like this)

PVince81 replied:
I've now used the binary form to detect all these, see last commit

juliushaertl commented on tests/lib/Security/Normalizer/IpAddressTest.php:
```diff
@@ -38,6 +38,10 @@ public function subnetDataProvider() {
 			],
+			[
+				'::ffff:192.168.0.123',
+				'::ffff:192.168.0.123/32',
```
I think the correct subnet mask for a IPv4 to IPv6 mapped address would be /128

PVince81 replied:
as discussed in chat we could agree to return the ipv4 form of subnet here, which would be "192.168.0.123/32"

PVince81 replied:
done in last commit

PVince81 committed:
Improve normalizer detecting IPv4 inside of IPv6

The subnet for an IPv4 address inside of IPv6 is now returned in its
IPv4 form.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

come-nc commented on lib/private/Security/Normalizer/IpAddress.php:
```diff
@@ -92,6 +92,39 @@ private function getIPv6Subnet(string $ip, int $maskBits = 48): string {
+	 * @param string $ipv6 IPv6 string
+	 * @return embedded IPv4 string or false if none was found
+	 */
+	private function getEmbeddedIpv4($ipv6) {
```
Since this is a new function please strong type instead of docblock

come-nc replied:
(only applicable for the parameter here, since return type is a union and we need to support PHP7)

come-nc commented on lib/private/Security/Normalizer/IpAddress.php:
```diff
@@ -92,6 +92,39 @@ private function getIPv6Subnet(string $ip, int $maskBits = 48): string {
+	 * The detected format is "::ffff:x.x.x.x" using the binary form.
+	 *
+	 * @param string $ipv6 IPv6 string
+	 * @return embedded IPv4 string or false if none was found
```
```suggestion
	 * @return false|string embedded IPv4 string or false if none was found
```

come-nc reviewed (changes requested):
Please fix the type declarations, it should please psalm as well.

juliushaertl commented on lib/private/Security/Normalizer/IpAddress.php:
```diff
@@ -92,6 +92,39 @@ private function getIPv6Subnet(string $ip, int $maskBits = 48): string {
+	 * The detected format is "::ffff:x.x.x.x" using the binary form.
+	 *
+	 * @param string $ipv6 IPv6 string
+	 * @return embedded IPv4 string or false if none was found
```
```suggestion
	 * @return string|false embedded IPv4 string or false if none was found
```

Or a bit cleaner maybe a nullable string instead of the mixed type

juliushaertl replied:
Seems Côme was faster

juliusknorr reviewed (approved):
Looks good apart from the phpdoc issue that psalm has. LGTM once that is resolved

come-nc commented on lib/private/Security/Normalizer/IpAddress.php:
```diff
@@ -92,6 +92,39 @@ private function getIPv6Subnet(string $ip, int $maskBits = 48): string {
+	 * Returns the IPv4 address embedded in an IPv6 if applicable.
+	 * The detected format is "::ffff:x.x.x.x" using the binary form.
+	 *
+	 * @param string $ipv6 IPv6 string
+	 * @return null|string embedded IPv4 string or null if none was found
+	 */
+	private function getEmbeddedIpv4($ipv6) {
+		$binary = inet_pton($ipv6);
```
```suggestion
	 * @return string|null embedded IPv4 string or null if none was found
	 */
	private function getEmbeddedIpv4(string $ipv6): ?string {
		$binary = inet_pton($ipv6);
```

PVince81 committed:
Type hint in IpAddress

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

Co-authored-by: Côme Chilliet <91878298+come-nc@users.noreply.github.com>

PVince81 commented:
failing test "log in with invalid user once fixed by admin" might be related: `/drone/src/tests/acceptance/features/login.feature:38` 

I'll investigate

PVince81 commented:
the test passes locally with docker, I wonder if it's timing related.
I'll restart CI just in case

PVince81 commented:
/backport to stable23

PVince81 commented:
/backport to stable22

PVince81 commented:
/backport to stable21
